cumulus_etl/fhir/fhir_schemas.py

```diff
@@ -3,16 +3,27 @@
 from collections import namedtuple
 from functools import partial
 
+import pyarrow
 from fhirclient.models import fhirabstractbase, fhirdate, fhirelementfactory
-import pyspark
 
 
 FhirProperty = namedtuple("FhirProperty", ["name", "json_name", "pytype", "is_list", "of_many", "required"])
 
 
-def create_spark_schema_for_resource(resource_type: str) -> pyspark.sql.types.StructType:
+def pyarrow_schema_from_resource_batch(resource_type: str, batch: list[dict]) -> pyarrow.Schema:
+    # Switch batch from a list of rows to a list of columns (which lets pyarrow infer full unionized struct types
+    # across all column values -- it seems to normally only want to infer from the first row if you use any of its
+    # row-oriented APIs, but if you give it columns, it looks at the whole set).
+    column_names = get_all_column_names(resource_type)
+    columns = {col_name: pyarrow.array(row.get(col_name) for row in batch) for col_name in column_names}
+    inferred_schema = pyarrow.struct({name: array.type for name, array in columns.items()})
+
+    return create_pyarrow_schema_for_resource(resource_type, inferred_schema)
```

Comment:
is it possible to create a schema from the expected FHIR datatype without inspecting the batch?

Reply:
Yes - and that's basically what we did before (the "wide" version).

But this code is adding the feature of "wide plus any deep fields in the batch" which requires inspecting the batch. The quick and dumb way I did was to use pyarrow's inferring support, but for that to work, it needs to examine columnar data.

Instead, what I could do is write some code that inspects every row/column of the batch and discovers the maximal "shape" of the structs values for those columns, so we know what fields to include in the final schema. Which is basically what pyarrow is doing. But with some memory cost.

Reply:
I ended up writing that code that inspects the shape manually (rather than converting the row-oriented data to column-oriented data). It wasn't so bad. Look for `merge_shape_of_dict()` in `fhir_schemas.py`